### PR TITLE
feat(extras): add `overseer.nvim` plugin extra

### PR DESCRIPTION
**Pull request by 0b10000000011**

### CODE REVIEW

File: lua/lazyvim/plugins/extras/overseer/core.lua [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        ft = "OverseerList",
+        open = function()
+          require("overseer").open()
+        end,
+      })
+    end,
+  },
+  {
+    "stevearc/overseer.nvim",
+    lazy = true,
```
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/overseer/core.lua:59]: Since you're defining `cmd` to lazy-load, maybe `lazy = true` is not needed?

Comment (SR-MyStar) [reply lua/lazyvim/plugins/extras/overseer/core.lua:59]: Maybe sure.
The left window removed `lazy = true,` line, right window not removed
图片
They all load `overseer.nvim` plugin on neovim open

Comment (SR-MyStar) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: Removed `lazy = true` line

Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: It's being loaded as a lualine dependency, that's why.

Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: You could maybe use `cond` in lualine to only show it when it's loaded, something like https://github.com/LazyVim/LazyVim/blob/a50f92f7550fb6e9f21c0852e6cb190e6fcd50f5/lua/lazyvim/plugins/ui.lua#L158-L161

Comment (SR-MyStar) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: Oh, I know.
On testd again, they all not load on neovim open, so `lazy = true` line is not needed.
图片

File: lua/lazyvim/plugins/extras/overseer/core.lua [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      },
+    },
+  },
+  {
+    "nvim-lualine/lualine.nvim",
+    optional = true,
+    opts = function(_, opts)
+      local overseer = require("overseer")
+      local lazyvim = require("lazy.core.config").spec.plugins["LazyVim"]
+      local icons = require("lazy.core.plugin").values(lazyvim, "opts", false).icons.overseer
```
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/overseer/core.lua:24]: I get an error on line 24
```
   Error  01:38:42 PM notify.error lazy.nvim Failed to run `config` for lualine.nvim

...ins/LazyVim/lua/lazyvim/plugins/extras/overseer/core.lua:24: attempt to index field 'icons' (a nil value)

# stacktrace:
  - lua/lazyvim/plugins/extras/overseer/core.lua:24 _in_ **values**
```
My workaround was to remove lines 23-24 and replace them with 
```lua
local icons = require("lazyvim.config").icons
```
Do you get the same error?

**Review by dpetka2001 — CHANGES_REQUESTED** [review]

File: lua/lazyvim/plugins/extras/overseer/core.lua (hunk 1 above)
Comment (SR-MyStar) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: > You could maybe use `cond` in lualine to only show it when it's loaded, something like
> 
> https://github.com/LazyVim/LazyVim/blob/a50f92f7550fb6e9f21c0852e6cb190e6fcd50f5/lua/lazyvim/plugins/ui.lua#L158-L161

Now it only show it in lualine when it's loaded

File: lua/lazyvim/plugins/extras/overseer/core.lua (hunk 2 above)
Comment (SR-MyStar) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: No, but I submited a patch to fix it

**Review by dpetka2001 — APPROVED** [review]

File: lua/lazyvim/plugins/extras/overseer/core.lua [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+              unique = false, -- Unique-ify non-running task count by name
+              name = nil, -- List of task names to search for
+              name_not = false, -- When true, invert the name search
+              status = nil, -- List of task statuses to display
+              status_not = false, -- When true, invert the status search
+            }
+          end,
+          cond = function()
+            return package.loaded["overseer"]
+          end,
```
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/overseer/core.lua:46]: I believe these lines should be inside the table that is returned as a lualine component, so right after `status_not = false` should be ok. Otherwise, I get a warning in `:LualineNotices`
```
Only functions, strings and tables can be used as component.
You seem to have a `table` as component indexed as `cond`.
Something like:

    cond = { <function 1> },

```

**Review by dpetka2001 — CHANGES_REQUESTED** [review]

File: lua/lazyvim/plugins/extras/overseer/core.lua (hunk 3 above)
Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: You should keep your first approach of using `opts` as a function and use `table.insert` to add the component to `lualine_x`, because the way you have it now will overwrite the default Lazyvim `lualine_x` component.

Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: Well what do you see in your lualine? With this change I can't see the `lazy.nvim`'s updates nor the `diff` component which shows if something was modified or deleted or added. Also for `opts`, `vim.tbl_deep_extend` is being used to recursively merge tables, but in this case the table is `opts` with subtables of `sections` and `lualine_x`. So it is essentially `opts.sections.lualine_x`, but `lualine_x` is a list and not a `key-value` table, so `vim.tbl_deep_extend` can't merge this and just overwrites it.

Comment (SR-MyStar) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: Sorry, I only tested it in my lualine config, Now I retested on after I removed my lualine config, you're right.

**Comment by SR-MyStar** [comment]
It's all wrong! I reviewed my config, lualine loaded old config file for not used `cond`, lualine module using `cond` has never show, I've been testing the old config!!!
Now I rollback this patch to the version for unused `cond`

**Comment by Jomik** [comment]
Did you give up on this one @SR-MyStar ?

**Review by Jomik — COMMENTED** [review]
I think this is missing `optional = true` for the neotest, dap and toggleterm additions.

**Comment by dpetka2001** [comment]
`optional = true` will only work for neotest and dap but not toggleterm, because toggleterm is listed as a dependency of overseer. See also [this issue on lazy.nvim repo](https://github.com/folke/lazy.nvim/issues/1402). I think it's related.

**Comment by Jomik** [comment]
> `optional = true` will only work for neotest and dap but not toggleterm, because toggleterm is listed as a dependency of overseer. See also [this issue on lazy.nvim repo](https://github.com/folke/lazy.nvim/issues/1402). I think it's related.

Yeah, but don't think toggleterm should be added as a required dependency.
It's optional.

**Comment by dpetka2001** [comment]
I agree with you, I just pointed out that it won't work since it's listed as a dependency in overseer because of the issue in `lazy.nvim`. If you want `optional` to work for toggleterm in the current state of things in `lazy.nvim`, it should be removed as a dependency to toggleterm. Maybe some other way of restructuring this PR regarding toggleterm and overseer would be better. I don't use overseer so I don't know, but is it necessary for toggleterm to be a dependency of it?

**Comment by Jomik** [comment]
You can have a `toggleterm` extra.
Then this is how we normally add support for plugins across.
```lua
return {
  {
    "akinsho/toggleterm.nvim",
    version = "*",
    opts = {},
  },
  {
    "stevearc/overseer.nvim",
    optional = true,
    dependencies = "akinsho/toggleterm.nvim",
    opts = {
      strategy = { "toggleterm" },
    },
  },
}
```

**Comment by dpetka2001** [comment]
Yes, this can also be an option, but from reading the overseer docs, I still don't see any reference of toggleterm having to be a dependency. You can just have the toggleterm extra like you just said without the dependency I believe. If the toggleterm extra is enabled then overseer will use `strategy = { "toggleterm" }`. Except if I'm missing something. But like I said I don't use overseer, so feel free to create an appropriate PR for this if you're interested.

**Comment by SR-MyStar** [comment]
> Did you give up on this one @SR-MyStar ?

No, but after I send it.

> I'm wrong! I reviewed my config, lualine loaded old config file for not used `cond`, lualine module using `cond` has never show, I've been testing the old config!!!
> 
> Now I rollback this patch to the version for unused `cond`.

Many days I'm not receive any notify, so before I'm not care it pull request, until now, I found your notify at google mail.

**Comment by SR-MyStar** [comment]
> Yes, this can also be an option, but from reading the overseer docs, I still don't see any reference of toggleterm having to be a dependency. You can just have the toggleterm extra like you just said without the dependency I believe. If the toggleterm extra is enabled then overseer will use `strategy = { "toggleterm" }`. Except if I'm missing something. But like I said I don't use overseer, so feel free to create an appropriate PR for this if you're interested.
> 
> PS: and please keep in mind that I'm just a regular user like you and these thoughts are purely personal based on my experience with LazyVim thus far.

You are right, I will fix it

**Comment by SR-MyStar** [comment]
Could I merge `nvim-dap` `neotest` `toggleterm` and `core` into a new file?

**Comment by kevintraver** [comment]
You could also add **Edgy** integration:

```
{
  "folke/edgy.nvim",
  optional = true,
  opts = function(_, opts)
    opts.bottom = opts.bottom or {}
    table.insert(opts.bottom, {
      ft = "OverseerList",
      title = "Overseer",
      size = { height = 0.4 },
    })
  end,
}
```

File: lua/lazyvim/plugins/extras/overseer/nvim-dap.lua [hunk 4]
```diff
@@ -0,0 +1,8 @@
+return {
+  "mfussenegger/nvim-dap",
+  dependencies = "stevearc/overseer.nvim",
+  opts = function()
+    require("overseer").patch_dap(true)
```
Comment (moetayuko) [inline lua/lazyvim/plugins/extras/overseer/nvim-dap.lua:5]: deprecated, check https://github.com/stevearc/overseer.nvim/blob/master/doc/third_party.md#dap

File: lua/lazyvim/plugins/extras/overseer/nvim-dap.lua [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+    require("dap.ext.vscode").json_decode = require("overseer.json").decode
```
Comment (moetayuko) [inline lua/lazyvim/plugins/extras/overseer/nvim-dap.lua:6]: Is this necessary when we already override its default json parser? https://github.com/LazyVim/LazyVim/blob/74c1766479411357054819b5d8c8376ca12c2504/lua/lazyvim/plugins/extras/dap/core.lua#L88-L93

**Comment by 0b10000000011** [comment]
I removed Lualine integration, because it will broke `overseer.nvim` lazy load mode and I don't know how to solve it.

**Comment by 0b10000000011** [comment]
Emmmm, I don't know how to solve it error.
```
 ./lua/lazyvim/plugins/extras/overseer/neotest.lua:6: module 'neotest.consumers.overseer' not found:
	no field package.preload['neotest.consumers.overseer']
cache_loader: module neotest.consumers.overseer not found
cache_loader_lib: module neotest.consumers.overseer not found
	no file './src/neotest/consumers/overseer.lua'
	no file './src/neotest/consumers/overseer/neotest/consumers/overseer.lua'
	no file './src/neotest/consumers/overseer/init.lua'
	no file './neotest/consumers/overseer.lua'
	no file '/home/runner/work/neovim/neovim/.deps/usr/share/luajit-2.1/neotest/consumers/overseer.lua'
	no file '/usr/local/share/lua/5.1/neotest/consumers/overseer.lua'
	no file '/usr/local/share/lua/5.1/neotest/consumers/overseer/init.lua'
	no file '/home/runner/work/neovim/neovim/.deps/usr/share/lua/5.1/neotest/consumers/overseer.lua'
	no file '/home/runner/work/neovim/neovim/.deps/usr/share/lua/5.1/neotest/consumers/overseer/init.lua'
	no file '/home/runner/work/LazyVim/LazyVim/.tests/data/nvim/lazy-rocks/busted/share/lua/5.1/neotest/consumers/overseer.lua'
	no file '/home/runner/work/LazyVim/LazyVim/.tests/data/nvim/lazy-rocks/busted/share/lua/5.1/neotest/consumers/overseer/init.lua'
	no file './csrc/neotest/consumers/overseer.so'
	no file './csrc/neotest/consumers/overseer/neotest/consumers/overseer.so'
	no file './neotest/consumers/overseer.so'
	no file '/usr/local/lib/lua/5.1/neotest/consumers/overseer.so'
	no file '/home/runner/work/neovim/neovim/.deps/usr/lib/lua/5.1/neotest/consumers/overseer.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
	no file '/home/runner/work/LazyVim/LazyVim/.tests/data/nvim/lazy-rocks/busted/lib/lua/5.1/neotest/consumers/overseer.so'
	no file './csrc/neotest.so'
	no file './csrc/neotest/neotest.so'
	no file './neotest.so'
	no file '/usr/local/lib/lua/5.1/neotest.so'
	no file '/home/runner/work/neovim/neovim/.deps/usr/lib/lua/5.1/neotest.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
	no file '/home/runner/work/LazyVim/LazyVim/.tests/data/nvim/lazy-rocks/busted/lib/lua/5.1/neotest.so'
```
On my host, it only show on the first open, so I think it's a priority question, but it show after I add it too.
https://github.com/LazyVim/LazyVim/blob/ed25852061ddb3664c285c1ed2bdb32e3a52d31f/lua/lazyvim/plugins/xtras.lua#L9

File: lua/lazyvim/plugins/extras/overseer/core.lua [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      "OverseerDeleteBundle",
+      "OverseerRunCmd",
+      "OverseerRun",
+      "OverseerInfo",
+      "OverseerBuild",
+      "OverseerQuickAction",
+      "OverseerTaskAction",
+      "OverseerClearCache",
+    },
+    dependencies = "nvim-telescope/telescope.nvim",
```
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/overseer/core.lua:35]: I don't believe this should be a hard dependency?

File: lua/lazyvim/plugins/xtras.lua [hunk 7]
```diff
@@ -6,6 +6,7 @@ local prios = {
   ["lazyvim.plugins.extras.lang.typescript"] = 5,
   ["lazyvim.plugins.extras.formatting.prettier"] = 10,
   -- default priority is 50
+  ["lazyvim.plugins.extras.overseer.neotest"] = 100,
```
Comment (dpetka2001) [inline lua/lazyvim/plugins/xtras.lua:9]: Why is this needed?

Comment (0b10000000011) [reply lua/lazyvim/plugins/xtras.lua]: I guess the prioritization of the error makes the error happen, so I write it and push on github to get CI outcome, it's not a right code, I will removed it now.

File: lua/lazyvim/plugins/extras/overseer/core.lua (hunk 6 above)
Comment (0b10000000011) [reply lua/lazyvim/plugins/extras/overseer/core.lua]: You are right

**Comment by dpetka2001** [comment]
Also could you please consolidate everything into one file? It doesn't make sense to have different overseer Extras. And just leave `toggleterm` out of the PR completely. It should be made into a separate PR of its own, although in my personal opinion since there isn't any configuration about it, it should be left to the user to install it in his personal configuration if he wants to.

**Comment by 0b10000000011** [comment]
Emmmmm, In `coding` or `editor`? (overseer.lua file)

**Comment by 0b10000000011** [comment]
> Also could you please consolidate everything into one file? It doesn't make sense to have different overseer Extras. And just leave `toggleterm` out of the PR completely. It should be made into a separate PR of its own, although in my personal opinion since there isn't any configuration about it, it should be left to the user to install it in his personal configuration if he wants to.

OK.

File: lua/lazyvim/plugins/extras/editor/overseer.lua [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        opts.right = opts.right or {}
+        table.insert(opts.right, {
+          title = "Overseer",
+          ft = "OverseerList",
+          open = function()
+            require("overseer").open()
+          end,
+        })
+      end,
+    },
```
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/editor/overseer.lua:48]: Ehm, this should not be inside the overseer spec?

File: lua/lazyvim/plugins/extras/editor/overseer.lua [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    opts = {
```
Comment (folke) [inline lua/lazyvim/plugins/extras/editor/overseer.lua:49]: this and keys shold be under the overseer.nvim plugin

File: lua/lazyvim/plugins/extras/editor/overseer.lua [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+      { "<leader>oi", "<cmd>OverseerInfo<cr>",        desc = "Overseer Info" },
+      { "<leader>ob", "<cmd>OverseerBuild<cr>",       desc = "Task builder" },
+      { "<leader>ot", "<cmd>OverseerTaskAction<cr>",  desc = "Task action" },
+      { "<leader>oc", "<cmd>OverseerClearCache<cr>",  desc = "Clear cache" },
+    },
+  },
+  {
+    "nvim-neotest/neotest",
+    optional = true,
+    dependencies = "stevearc/overseer.nvim",
```
Comment (folke) [inline lua/lazyvim/plugins/extras/editor/overseer.lua:89]: This is not needed

File: lua/lazyvim/plugins/extras/editor/overseer.lua [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    opts = {
+      consumers = {
+        overseer = require("neotest.consumers.overseer"),
```
Comment (folke) [inline lua/lazyvim/plugins/extras/editor/overseer.lua:92]: This will always load neotest & overseer during spec parsing. 
Please put it in a closure

File: lua/lazyvim/plugins/extras/editor/overseer.lua (hunk 10 above)
Comment (folke) [reply lua/lazyvim/plugins/extras/editor/overseer.lua:89]: I meant the `dependencies` line

File: lua/lazyvim/plugins/extras/editor/overseer.lua (hunk 8 above)
Comment (0b10000000011) [reply lua/lazyvim/plugins/extras/editor/overseer.lua]: Oh, sorry.

File: lua/lazyvim/plugins/extras/editor/overseer.lua (hunk 11 above)
Comment (0b10000000011) [reply lua/lazyvim/plugins/extras/editor/overseer.lua:92]: This?
```lua
{
  "nvim-neotest/neotest",
  optional = true,
  opts = function(_, opts)
    opts = opts or {}
    opts.consumers = opts.consumers or {}
    opts.consumers.overseer = require("neotest.consumers.overseer")
  end,
}
```

Comment (0b10000000011) [reply lua/lazyvim/plugins/extras/editor/overseer.lua]: Oh, I know.

File: lua/lazyvim/plugins/extras/editor/overseer.lua (hunk 10 above)
Comment (0b10000000011) [reply lua/lazyvim/plugins/extras/editor/overseer.lua]: OK, removed.

File: lua/lazyvim/plugins/extras/editor/overseer.lua (hunk 9 above)
Comment (0b10000000011) [reply lua/lazyvim/plugins/extras/editor/overseer.lua:26]: Already fixed.

**Comment by folke** [comment]
Cleaned up and merged. Thank you!
